File: r/robot-bounded-in-circle.rb
# @param {String} instructions
# @return {Boolean}

def is_robot_bounded(instructions)
  directions = %w[N E S W]
  # Corresponds to N, E, S, W resp
  current_coordinates = [0, 0, 0, 0]
  current_direction = 0

  instructions.chars.each do |instruction|
    case instruction
    when 'G'
      current_coordinates[current_direction] += 1
    when 'L'
      if current_direction == 0
        current_direction = 3
      else
        current_direction -= 1
      end
    when 'R'
      if current_direction == 3
        current_direction = 0
      else
        current_direction += 1
      end
    end
  end

  result_coordinates = [
    current_coordinates[0] - current_coordinates[2],
    current_coordinates[1] - current_coordinates[3]
  ]

  result_coordinates == [0, 0] || current_direction != 0
end

File: r/robot-bounded-in-circle.cc
const int dx[4] = {-1, 0, 1, 0}, dy[4] = {0, 1, 0, -1};

class Solution {
public:
  bool isRobotBounded(string instructions) {
    instructions += instructions;
    instructions += instructions;
    int dir = 0, cx = 0, cy = 0;
    for (char c : instructions) {
      if (c == 'G') cx += dx[dir], cy += dy[dir];
      else if (c == 'L') dir = (dir + 3) % 4;
      else dir = (dir + 1) % 4;
    }
    return cx == 0 && cy == 0;
  }
};
